vpn-slice: fix script startup with python > 3.11 on darwin, modernize a little

dwt opened the pull request:
## Things done

Python 3.12 removed distutils from the standard library, on which vpn-slice depends as a fallback if setuptools is not available.

In testing I tried adding distutils to nativeBuildInputs of buildInputs, but neithe worked (not sure why, this should not be needed at runtime, but is probably an interaction with buildPythonApplication).

Closes #385323

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dwt committed:
vpn-slice: fix script startup with python > 3.11, modernize a little

Python 3.12 removed distutils from the standard library, on which
vpn-slice depends as a fallback if setuptools is not available.

In testing I tried adding distutils to nativeBuildInputs of buildInputs,
but neithe worked (not sure why, this should not be needed at runtime,
but is probably an interaction with buildPythonApplication).

dwt commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 385356`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>vpn-slice</li>
    <li>vpn-slice.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>vpn-slice</li>
    <li>vpn-slice.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>vpn-slice</li>
    <li>vpn-slice.dist</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>vpn-slice</li>
    <li>vpn-slice.dist</li>
  </ul>
</details>

afh commented:
When building and running this locally I see the following error:

```sh
./result/bin/vpn-slice
No module named 'distutils'
```

Is there more that needs to be done in order to make vpn-slice work with setuptools instead of distutils?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/1016/189

afh reviewed (approved):
LGTM

9999years commented on pkgs/tools/networking/vpn-slice/default.nix:
```diff
@@ -35,6 +35,7 @@ buildPythonApplication rec {
   propagatedBuildInputs = with python3Packages; [
     setproctitle
     dnspython
+    setuptools
```
I think you could put this in the `build-system` argument instead, see docs here: https://nixos.org/manual/nixpkgs/unstable/#buildpythonpackage-parameters

dwt replied:
:-(

```diff
diff --git a/pkgs/tools/networking/vpn-slice/default.nix b/pkgs/tools/networking/vpn-slice/default.nix
index 047f3f0c7acd..2b464088b063 100644
--- a/pkgs/tools/networking/vpn-slice/default.nix
+++ b/pkgs/tools/networking/vpn-slice/default.nix
@@ -32,10 +32,13 @@ buildPythonApplication rec {
         --replace "'/sbin/iptables'" "'${iptables}/bin/iptables'"
     '';
 
+  build-system = with python3Packages; [
+    setuptools
+  ];
+
   propagatedBuildInputs = with python3Packages; [
     setproctitle
     dnspython
-    setuptools
   ];
 
   doCheck = false;
```

Does in fact not work.

9999years replied:
Very odd, thanks for trying it!

SuperSandro2000 replied:
You must set pyproject = true for this to work

dwt replied:
```diff
diff --git a/pkgs/tools/networking/vpn-slice/default.nix b/pkgs/tools/networking/vpn-slice/default.nix
index 047f3f0c7acd..4551d5fd212f 100644
--- a/pkgs/tools/networking/vpn-slice/default.nix
+++ b/pkgs/tools/networking/vpn-slice/default.nix
@@ -32,10 +32,15 @@ buildPythonApplication rec {
         --replace "'/sbin/iptables'" "'${iptables}/bin/iptables'"
     '';
 
+  pyproject = true;
+  build-system = with python3Packages; [
+    setuptools
+  ];
+
   propagatedBuildInputs = with python3Packages; [
     setproctitle
     dnspython
-    setuptools
+    # setuptools
   ];
 
   doCheck = false;
```

This also doesn't work. @SuperSandro2000 can you perhaps point me to some documentation that explains how this is supposed to work? The Nixpkgs documentation is a bit terse on details for me.

SuperSandro2000 replied:
You also need to change propagatedBuildInputs to dependencies. see https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonpackage-parameters-buildpythonpackage-parameters

I just pushed the change myself.

dwt commented:
@9999years and @SuperSandro2000 Do you approve these changes, or would you need additional changes?

9999years commented:
@dwt LGTM

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2291

dwt commented:
Uhm.... @SuperSandro2000 how did you check the changes you force pushed last?

Because the change to `pyproject = true` and pulling the `setuptools` dependency out to the `build-system` does not work for me on darwin. I would like to know how it can work for you?

SuperSandro2000 commented:
nix build .#vpn-slice for x86_64-linux and aarch64-linux

dwt commented:
@SuperSandro2000 I have reverted the part of the change that added the `build-system` and added a comment that explains why it's no good (as it should be tempting for everyone who understands nix python build processes to repeat that error).

SuperSandro2000 commented:
> It should fail because it cannot write to /etc/hosts (good) or because it cannot import distutils (bad, missing setuptools dependency).

I tried exactly that. I cannot reproduce it. Please post your entire logs so that we can debug this

SuperSandro2000 commented on pkgs/tools/networking/vpn-slice/default.nix:
```diff
@@ -13,26 +13,38 @@
     '';
 
-  propagatedBuildInputs = with python3Packages; [
+  # One would expect setuptools should be part of the build-system
```
Please remove this entire comment. What you are experience just means that the tool also has a runtime dependency on setuptools which means we need it in both which is totally valid in the python ecosystem.

Just according to the source code, I couldn't find any reference to setuptools besides the setup.py file which is not needed at runtime. Where is that actually required? What is the error you

dwt replied:
[x] done

SuperSandro2000 commented on pkgs/tools/networking/vpn-slice/default.nix:
```diff
@@ -13,26 +13,38 @@
+  # to test, call the built `vpn-slice` without arguments,
+  # it should fail because it cannot write to /etc/hosts (good)
+  # not with the package `distutils` missing (it internally falls
+  # back from setuptools to distutils if setuptools is missing)
```
distuitls is deprecated and removed in python 3.13. there is a backport package but we should try to avoid that, too if possible.

dwt replied:
It was already removed in python 3.12

> distutils — Building and installing Python modules
> Deprecated since version 3.10, removed in version 3.12.

See https://docs.python.org/3.12/library/distutils.html#module-distutils

dwt replied:
[x] done

dwt commented:
@SuperSandro2000 

With 
```diff
diff --git a/pkgs/tools/networking/vpn-slice/default.nix b/pkgs/tools/networking/vpn-slice/default.nix
index 63bfe3067cfc..311f403c98bd 100644
--- a/pkgs/tools/networking/vpn-slice/default.nix
+++ b/pkgs/tools/networking/vpn-slice/default.nix
@@ -39,12 +39,12 @@ buildPythonApplication rec {
   # it should fail because it cannot write to /etc/hosts (good)
   # not with the package `distutils` missing (it internally falls
   # back from setuptools to distutils if setuptools is missing)
-  # build-system = with python3Packages; [
-  #  setuptools
-  # ];
+  build-system = with python3Packages; [
+    setuptools
+  ];
 
   dependencies = with python3Packages; [
-    setuptools
+    # setuptools
     setproctitle
     dnspython
   ];
```

```shell
❯ nix-build --attr vpn-slice
/nix/store/sgyp28j2vjyymw3jkjbx96i56iq8x7jy-vpn-slice-0.16.1
❯ result/bin/vpn-slice
No module named 'distutils'
# I have also tested using it in an actual vpn connection, same problem
```

with this `1dd3dfc1` commit, the result is:

```shell
❯ nix-build --attr vpn-slice
/nix/store/yya6i39vb8paxhwidww7dwv7xj5cihn1-vpn-slice-0.16.1
❯ result/bin/vpn-slice
WARNING: Couldn't configure hosts provider: Cannot read/write /etc/hosts
Aborting because providers for hosts are required; use --help for more information
```

Just to be sure, the vpn-slice script should build against python3.12

```shell
❯ head -1 /nix/store/yya6i39vb8paxhwidww7dwv7xj5cihn1-vpn-slice-0.16.1/bin/.vpn-slice-wrapped
#!/nix/store/wwqdmdr2f5wrjnsjs64bny8df471rh9b-python3-3.12.9/bin/python3.12
```
as older versions of python will still have `distutils` in the sodlib.

SuperSandro2000 commented:
Just place setuptools then in both. That isn't a problem. I just don't understand why we don't get a backtrace and see where the import is failing. Probably upstream is masking the error....

And we don't want to rely on an older python version ideally either. At some point dependencies also won't work with that.

dwt commented:
Setting `PYTHONVERBOSE=x` gives some hints, this is where it crashes:

```logs
# code object from '/nix/store/sgyp28j2vjyymw3jkjbx96i56iq8x7jy-vpn-slice-0.16.1/lib/python3.12/site-packages/vpn_slice/__pycache__/provider.cpython-312.pyc'
import 'vpn_slice.provider' # <_frozen_importlib_external.SourceFileLoader object at 0x1075b0bf0>
import 'vpn_slice.dnspython' # <_frozen_importlib_external.SourceFileLoader object at 0x10496a8a0>
No module named 'distutils'
```

dwt commented:
@SuperSandro2000 [I guess this check for darwin is the culprit](https://github.com/dlenski/vpn-slice/blob/d1f419ad20b5e2590222a93dfa6530e6d91701a6/vpn_slice/__main__.py#L44)

dwt commented:
@SuperSandro2000 is that good enough or do you need anything else to get this merged?

SuperSandro2000 commented:
> @SuperSandro2000 [I guess this check for darwin is the culprit](https://github.com/dlenski/vpn-slice/blob/d1f419ad20b5e2590222a93dfa6530e6d91701a6/vpn_slice/__main__.py#L44), then later [this eats the exception backtrace](https://github.com/dlenski/vpn-slice/blob/d1f419ad20b5e2590222a93dfa6530e6d91701a6/vpn_slice/__main__.py#L587). I'll open a bug report upstream.

sigh ....

https://github.com/dlenski/vpn-slice/commit/4e26adbfd14de2be5e77933e96d353ea7d200107

SuperSandro2000 commented:
Thanks for bearing with me and solving this properly :)

dwt commented:
Thanks!

And yes, I've added a comment that we can remove this runtime dependency with one of the next updates.

dwt commented:
You are welcome. :)
